src/expr.rs

```diff
@@ -35,7 +37,222 @@ pub enum Expr {
+        v
+    }
+}
+
+
+impl TryInto<f64> for Expr {
+    type Error = ();
+    fn try_into(self) -> Result<f64, Self::Error> {
+        match self {
+            Expr::Float(v) => Ok(v),
+            Expr::Literal(v) => f64::from_str(&v).or(Err(())),
+            _ => Err(()),
+        }
+    }
+}
+
+impl TryInto<bool> for Expr {
+    type Error = ();
+    fn try_into(self) -> Result<bool, Self::Error> {
+        match self {
+            Expr::Bool(v) => Ok(v),
+            _ => Err(()),
+        }
+    }
+}
+
 impl Expr {
+    /// Insert values from properties from json
+    ///
+    ///  # Examples
+    ///
+    /// ```
+    /// use serde_json::{json, Value};
+    /// use cql2::Expr;
+    /// let item = json!({"properties":{"eo:cloud_cover":10, "datetime": "2020-01-01 00:00:00Z", "boolfield": true}});
+    /// let mut expr_json = json!(
+    ///     {
+    ///         "op": "+",
+    ///         "args": [
+    ///             {"property": "eo:cloud_cover"},
+    ///             10
+    ///         ]
+    ///     }
+    /// );
+    ///
+    /// let mut expr: Expr = serde_json::from_value(expr_json).unwrap();
+    /// println!("Initial {:?}", expr);
+    /// expr.reduce(&item);
+    ///
+    /// let output: f64;
+    /// if let Expr::Float(v) = expr {
+    ///     output = v;
+    /// } else {
+    ///     assert!(false);
+    ///     output = 0.0;
+    /// }
+    /// println!("Modified {:?}", expr);
+    ///
+    /// assert_eq!(20.0, output);
+    ///
+    ///
+    /// ```
+    pub fn reduce(&mut self, j: &Value) {
+        match self {
+            Expr::Property { property } => {
+                let mut prefixproperty: String = "properties.".to_string();
+                prefixproperty.push_str(property);
+
+                let mut propexpr: Option<Value> = None;
+                if j.dot_has(property) {
+                    propexpr = j.dot_get(property).unwrap();
+                } else {
+                    let mut prefixproperty: String = "properties.".to_string();
+                    prefixproperty.push_str(property);
+                    propexpr = j.dot_get(&prefixproperty).unwrap();
+                }
+                if let Some(v) = propexpr {
+                    *self = Expr::from(v);
+                }
+            }
+            Expr::Operation { op, args } => {
+                let mut alltrue: bool = true;
+                let mut anytrue: bool = false;
+                let mut allbool: bool = true;
+                for arg in args.iter_mut() {
+                    arg.reduce(j);
+                    let b: Result<bool, _> = arg.as_ref().clone().try_into();
+                    match b {
+                        Ok(true) => anytrue = true,
+                        Ok(false) => {
+                            alltrue = false;
+                        }
+                        _ => {
+                            alltrue = false;
+                            allbool = false;
+                        }
+                    }
```

Comment:
Not sure about the `.as_ref().clone()` removal, but I'd guess there'd be a way to avoid that? I'm working in the browser right now so don't have IDE powers to make me less dumb 😄:

```suggestion
                    if let Ok(bool) = arg.try_into() {
                        if bool {
                            anytrue = true;
                        } else {
                            alltrue = false;
                        }
                    } else {
                        alltrue = false;
                        allbool = false;
                        // Can you break here?
                    }
```

Reply:
Can't break there as the main thing that loop is doing is to run reduce against each arg. We are just doing the bool check as an addition so we only have to loop through the args once.